stock_available_to_promise_release/models/stock_move.py

```diff
@@ -591,6 +597,64 @@ def _get_chained_moves_iterator(self, chain_field):
             yield moves
             moves = moves.mapped(chain_field)
 
+    def _return_quantity_in_stock(self, moves_to_return_per_type):
+        """Return a quantity from a list of moves.
+
+        The quantity to return is in the product uom"""
+        for picking_type, qty_per_move in moves_to_return_per_type.items():
+            moves = self.browse()
+            for move in qty_per_move:
+                moves |= move
+            picking = fields.first(moves.picking_id)
+            if not picking:
+                continue
+            return_type = picking_type.return_picking_type_id
+            wiz_values = {
+                "picking_id": picking.id,
+                "location_id": return_type.default_location_dest_id.id,
+            }
+            product_return_moves = []
+            if not return_type:
+                message = _(
+                    "The operation {picking_name} is done and cannot be returned",
+                    picking_name=picking.name,
+                )
+                raise UserError(message)
+            for move, qty_to_return in qty_per_move.items():
+                # Cannot return an unprocessed move
+                if move.state != "done":
+                    continue
+                product = move.product_id
+                uom = product.uom_id
+                # Cannot return 0 qty
+                if float_is_zero(qty_to_return, precision_rounding=uom.rounding):
+                    continue
+                return_move_vals = {
+                    "product_id": product.id,
+                    "quantity": qty_to_return,
+                    "uom_id": uom.id,
+                    "move_id": move.id,
+                }
+                product_return_moves.append((0, 0, return_move_vals))
+            if product_return_moves:
+                wiz_values["product_return_moves"] = product_return_moves
+                return_wiz = self.env["stock.return.picking"].create(wiz_values)
+                return_wiz.create_returns()
+        return True
+
+    def _get_returneable_qty_per_move(self, qty_to_return):
+        res = {}
+        for move in self.filtered(lambda m: m.state == "done"):
```

Comment:
you are already processing done moves
```suggestion
        for move in self:
```

Reply:
This method is helpful and could be used elsewhere, we need to add some restrictions.
Reusability at the cost of an extra check is good, IMO.
I'll drop this if you insist, but I kind of disagree.